File: lib/services/health_check_service.dart
import '../models/readme_element.dart';

enum IssueSeverity { error, warning, info, success }

class HealthIssue {
  final String message;
  final IssueSeverity severity;
  final String? elementId;
  final String? suggestion;

  HealthIssue({
    required this.message,
    required this.severity,
    this.elementId,
    this.suggestion,
  });
}

class HealthCheckService {
  static List<HealthIssue> analyze(List<ReadmeElement> elements) {
    final List<HealthIssue> issues = [];

    if (elements.isEmpty) {
      issues.add(HealthIssue(
        message: 'Your README is a blank canvas.',
        severity: IssueSeverity.info,
        suggestion: 'Start by adding a Project Title (H1) to introduce your work.',
      ));
      return issues;
    }

    // 1. Critical Structure Analysis
    bool hasH1 = false;
    int imageCount = 0;
    int linkCount = 0;

    for (final element in elements) {
      if (element is HeadingElement && element.level == 1) hasH1 = true;
      if (element is ImageElement) imageCount++;
      if (element is LinkButtonElement || element is BadgeElement) linkCount++;
    }

    if (!hasH1) {
      issues.add(HealthIssue(
        message: 'Missing Main Title.',
        severity: IssueSeverity.error,
        suggestion: 'Add an H1 heading with your project name for better SEO.',
      ));
    }

    // 2. Visual & Media Analysis
    if (imageCount == 0) {
      issues.add(HealthIssue(
        message: 'Low visual appeal.',
        severity: IssueSeverity.warning,
        suggestion: 'Consider adding a screenshot or a demo GIF to showcase your app.',
      ));
    }

    // 3. Social & Networking Analysis
    if (!elements.any((e) => e is SocialsElement)) {
      issues.add(HealthIssue(
        message: 'Community links missing.',
        severity: IssueSeverity.info,
        suggestion: 'Add a Socials element so people can find and support you.',
      ));
    }

    // 4. Code Quality Check
    for (final element in elements) {
      if (element is CodeBlockElement && element.code.length < 10) {
        issues.add(HealthIssue(
          message: 'Short code block detected.',
          severity: IssueSeverity.warning,
          elementId: element.id,
          suggestion: 'Provide a meaningful code example or installation step.',
        ));
      }
    }

    // 5. Documentation Balance (Algo)
    if (elements.length > 3 && issues.isEmpty) {
      issues.add(HealthIssue(
        message: 'Documentation looking great!',
        severity: IssueSeverity.success,
        suggestion: 'You have a well-balanced structure. Ready to publish!',
      ));
    }

    return issues;
  }

  static double calculateDocumentationScore(List<ReadmeElement> elements) {
    if (elements.isEmpty) return 0;
    
    double score = 20; // Base score for starting
    
    // Structure points
    if (elements.any((e) => e is HeadingElement && e.level == 1)) score += 20;
    if (elements.any((e) => e is HeadingElement && e.level == 2)) score += 10;
    
    // Content points
    if (elements.any((e) => e is ParagraphElement)) score += 15;
    if (elements.any((e) => e is ImageElement)) score += 15;
    if (elements.any((e) => e is CodeBlockElement)) score += 10;
    if (elements.any((e) => e is ListElement)) score += 10;
    
    return score.clamp(0, 100);
  }
}

File: src/services/healthCheckService.js
// HealthCheckService (React Native)
// Analyzes README elements for issues
export const IssueSeverity = { error: 'error', warning: 'warning', info: 'info', success: 'success' };
export class HealthIssue {
  constructor({ message, severity, elementId, suggestion }) {
    this.message = message;
    this.severity = severity;
    this.elementId = elementId;
    this.suggestion = suggestion;
  }
}
export class HealthCheckService {
  static analyze(elements) {
    const issues = [];
    if (!elements || elements.length === 0) {
      issues.push(new HealthIssue({
        message: 'Your README is empty.',
        severity: IssueSeverity.info,
        suggestion: 'Add a Title (H1) to start your project documentation.'
      }));
      return issues;
    }
    let hasH1 = false;
    let hasDescription = false;
    let hasInstall = false;
    let imageCount = 0;
    for (const element of elements) {
      if (element.type === 'heading' && element.level === 1) hasH1 = true;
      if (element.type === 'paragraph' && element.text && element.text.length > 30) hasDescription = true;
      if (element.type === 'heading' && /install/i.test(element.text)) hasInstall = true;
      if (element.type === 'image') imageCount++;
      if (element.type === 'heading' && (!element.text || !element.text.trim())) {
        issues.push(new HealthIssue({ message: 'Empty heading detected.', severity: IssueSeverity.error, elementId: element.id }));
      }
      if (element.type === 'image' && (!element.altText || !element.altText.trim())) {
        issues.push(new HealthIssue({ message: 'Image missing alt text.', severity: IssueSeverity.warning, elementId: element.id, suggestion: 'Add alt text for accessibility.' }));
      }
      if ((element.type === 'linkButton' || element.type === 'button') && (!element.url || !element.url.trim())) {
        issues.push(new HealthIssue({ message: 'Button with no URL.', severity: IssueSeverity.error, elementId: element.id }));
      }
    }
    if (!hasH1) issues.push(new HealthIssue({ message: 'Missing H1 Title.', severity: IssueSeverity.error, suggestion: 'Add a main heading (H1).' }));
    if (!hasDescription) issues.push(new HealthIssue({ message: 'No project description.', severity: IssueSeverity.warning, suggestion: 'Add a paragraph describing your project.' }));
    if (!hasInstall) issues.push(new HealthIssue({ message: 'No installation section.', severity: IssueSeverity.info, suggestion: 'Add an Installation section.' }));
    if (imageCount === 0) issues.push(new HealthIssue({ message: 'No images found.', severity: IssueSeverity.info, suggestion: 'Add at least one image or badge.' }));
    return issues;
  }

  static calculateDocumentationScore(elements) {
    if (!elements || elements.length === 0) return 0;
    let score = 30;
    if (elements.some((e) => e.type === 'heading' && e.level === 1)) score += 20;
    if (elements.some((e) => e.type === 'paragraph' && (e.text || '').length > 50)) score += 20;
    if (elements.some((e) => e.type === 'image')) score += 10;
    if (elements.some((e) => e.type === 'code' || e.type === 'codeBlock')) score += 10;
    if (elements.some((e) => e.type === 'socials')) score += 10;
    return Math.min(100, Math.max(0, score));
  }
}
